pkg/apis/hive/v1alpha1/syncset_types.go

```diff
@@ -0,0 +1,213 @@
+
+	// UpdateSuccess indicates whether the resource has been updates or not. If not,
+	// it should include a reason and message for the failure.
+	UpdateSuccess SyncConditionType = "UpdateSuccess"
+
+	// PatchSuccess indicates whether the patch has been applied or not. If not,
+	// it should include a reason and message for the failure.
+	PatchSuccess SyncConditionType = "PatchSuccess"
+
+	// DeletionFailed indicates that resource deletion has failed. It should include
+	// a reason and message for the failure.
+	DeletionFailed SyncConditionType = "DeletionFailed"
+)
+
+// SyncCondition is a condition in a SyncStatus
+type SyncCondition struct {
+	// Type is the type of the condition.
+	Type SyncConditionType `json:"type"`
+	// Status is the status of the condition.
+	Status corev1.ConditionStatus `json:"status"`
+	// LastProbeTime is the last time we probed the condition.
+	// +optional
+	LastProbeTime metav1.Time `json:"lastProbeTime,omitempty"`
+	// LastTransitionTime is the last time the condition transitioned from one status to another.
+	// +optional
+	LastTransitionTime metav1.Time `json:"lastTransitionTime,omitempty"`
+	// Reason is a unique, one-word, CamelCase reason for the condition's last transition.
+	// +optional
+	Reason string `json:"reason,omitempty"`
+	// Message is a human-readable message indicating details about last transition.
+	// +optional
+	Message string `json:"message,omitempty"`
+}
+
+// SyncStatus describes objects that have been created or patches that
+// have been applied using the unique md5 sum of the object or patch.
+type SyncStatus struct {
+	GroupVersionKind schema.GroupVersionKind `json:"groupVersionKind"`
+	Name             string                  `json:"name"`
+	Namespace        string                  `json:"namespace"`
+	Hash             []byte                  `json:"hash"`
+	Conditions       []SyncCondition         `json:"conditions"`
+}
+
+// SyncSetCommonSpec defines the resources and patches to sync
+type SyncSetCommonSpec struct {
+	// Resources is the list of objects to sync.
+	// +optional
+	Resources []runtime.RawExtension `json:"resources,omitempty"`
+
+	// ResourceApplyMode indicates if the resource apply mode is "upsert" (default) or "sync".
+	// ApplyMode "upsert" indicates create and update.
+	// ApplyMode "sync" indicates create, update and delete.
+	// +optional
+	ResourceApplyMode SyncSetResourceApplyMode `json:"resourceApplyMode,omitempty"`
+
+	// Patches is the list of patches to apply.
+	// +optional
+	Patches []SyncObjectPatch `json:"patches,omitempty"`
+}
+
+// SelectorSyncSetSpec defines the SyncSetCommonSpec resources and patches to sync along
+// with a ClusterDeploymentSelector indicating which clusters the SelectorSyncSet applies
+// to in any namespace.
+type SelectorSyncSetSpec struct {
+	SyncSetCommonSpec         `json:",inline"`
+	ClusterDeploymentSelector metav1.LabelSelector `json:"clusterDeploymentSelector,omitempty"`
```

Comment:
If using the ` //+optional ` comment for all optional values, use it consistently
Add field comments